boardswarm-cli/src/ui.rs

```diff
@@ -50,21 +70,79 @@ impl Terminal {
         self.parser.set_scrollback(0)
     }
 
+    fn set_functioning(&mut self, str: String) {
+        self.terminal_mode = Some(TerminalMode::Functioning(str));
+        self.message = None;
+    }
+
+    fn set_non_functioning(&mut self, str: String) {
+        self.terminal_mode = Some(TerminalMode::NonFunctioning(str));
+        self.message = None;
+    }
+
+    fn set_alternate(&mut self, str: String) {
+        self.terminal_mode = Some(TerminalMode::Alternate(str));
+        self.message = None;
+    }
+
+    fn set_message(&mut self, message: String) {
+        self.message = Some(message);
+    }
+
     fn process(&mut self, bytes: &[u8]) {
         self.parser.process(bytes);
     }
 
     async fn update(&mut self) {
         let screen = self.parser.screen();
         let term = ui_term::UiTerm::new(screen);
+        let terminal_mode = self.terminal_mode.clone();
+        let message = self.message.clone();
         self.tui
             .draw(|f| {
                 let size = f.size();
-                let term_area = Rect::new(0, 0, 80.min(size.width), 24.min(size.height));
-                f.render_widget(term, term_area);
+                let term_area = Rect::new(0, 0, COLUMNS.min(size.width), ROWS.min(size.height));
+                let area_chunks = Layout::default()
+                    .direction(Direction::Vertical)
+                    .constraints([Constraint::Min(1), Constraint::Length(1)])
```

Comment:
I assume this means that if the terminal size is actually 80x24 you'll only see 23 lines? Would it make sense to add a min constraint of 24 for the term so we prioritize over showing that fully *over* showing a status bar?

Reply:
Yes, in the current shape of the series I do steal the lowest line for the status bar.

You guys settle on what you want first https://github.com/boardswarm/boardswarm/pull/90#discussion_r1622021635

Actually I imagine that you maybe _don't_ know and this thread is exactly about _learning_ what we want and I'm okay with that. Only I wouldn't like to go in circles too many times. ;P

It seems to me that relaxing the 80x24 is a bit of an orthogonal problem. IOW trying to conditionally display the status bar (sometimes display, sometimes not) is effectively trying to work around not being able to use other terminal sizes. Then there's a question of _when_ to display and when not. Then the user might want to recall what the current mode is. It's already been tried in this thread. Displaying the status bar all the time does simplify the code.

Reply:
I don't need to settle on anything really :) Displaying the console is the obvious primary function of the UI everything else is secondary (hence that's all it does at the moment); We shouldn't ever force the terminal area to be smaller then what it wants to be as that will likely hide output lines; e.g. if the last line is hidden your shell input will disappear quite quickly; if the top line is hidden you're missing an important part of e.g. `top`.

Not being able to set the size is orthogonal to this and primarily it's not really a boardswarm client call but depends on what the remote console side thinks it's rendering to, so we can't just decide "relax" and change size because we want to fit a statusbar :) Imho this should go as far as the UI displaying an error if the output size is smaller then its terminal size, because, really, we should never hide things in the terminal as that quickly ends up confusing. 

For 80x24, that's still the default when opening most terminal emulators these days even if it's somewhat historical. And similar a serial tty on linux/u-boot serial output etc all also start at 80x24, hence the hardcoding in the UI at the moment. But that also fundamentally means we should support displaying 80x24  on a 80x24, hence dropping out the status bar in those cases to not be annoying in that situation :)